### PR TITLE
Select producers revisit

### PR DESCRIPTION
**Pull request by engwarrior**
https://www.notion.so/maticnetwork/Slot-selection-612aa392286f4d0f8bebc7e3beafddb8 

Remove shuffle in favour of much simpler validators random algorithm selection based on powers.

Passes unit tests.

### CODE REVIEW

File: bor/selection.go [hunk 1]
```diff
@@ -1,13 +1,31 @@
...
+	"encoding/binary"
+	"math/rand"
 
-	"github.com/maticnetwork/heimdall/bor/types"
+	"github.com/maticnetwork/bor/common"
 	"github.com/maticnetwork/heimdall/helper"
 	hmTypes "github.com/maticnetwork/heimdall/types"
 )
 
+func binarySearch(array []uint64, search uint64) int {
```
[the remark this answers is not in the document]
Comment (engwarrior) [reply bor/selection.go:12]: I think to utils maybe?

File: bor/selection.go [hunk 2]
```diff
@@ -18,23 +36,27 @@ func SelectNextProducers(blkHash common.Hash, spanEligibleVals []hmTypes.Validat
...
 
-// converts validator power to slots
-// TODO remove 2nd loop
-func convertToSlots(vals []hmTypes.Validator) (validatorIndices []uint64) {
-	for _, val := range vals {
-		for val.VotingPower >= types.SlotCost {
-			validatorIndices = append(validatorIndices, uint64(val.ID))
-			val.VotingPower = val.VotingPower - types.SlotCost
+// makes ranges out of validators powers
+func convertToRanges(vals []hmTypes.Validator) []uint64 {
```
[the remark this answers is not in the document]
Comment (engwarrior) [reply bor/selection.go:52]: Isn't it overkill? Maybe I'll just add broad description? It just calculates ranges based on staking powers. (prefix sums)

[the remark this answers is not in the document]
Comment (tokamak-git) [reply bor/selection.go:52]: just a few cases would suffice, helps keep coverage high and reduces testing debt for the high level tests don't you think?

File: bor/selection.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tokamak-git) [reply bor/selection.go:12]: personally I don't like `util` packages, they end up being a dumping ground. that's why I created a file package for file related functionality. maybe @jdkanani or @mankenavenkatesh can weigh in on this...

**Comment by vaibhavchellani** [comment]
Would really appreciate more testcases on this as this is a very crucial component, any bias towards any validator will harm the network. 

The current testcases work with a very small dataset and apart from the last 2 testcases others don't even hit the `selection` algorithm.

**Pull request closed without merging**